mint-auditor/src/gnosis/config.rs

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! Configuration for Gnosis safe auditing
+
+use super::{Error, EthAddr};
+use mc_transaction_core::TokenId;
+use serde::{Deserialize, Serialize};
+use std::{fs, path::Path};
+use url::Url;
+
+/// Configuration for a token we want to audit.
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
+pub struct AuditedToken {
+    /// The MobileCoin token id.
+    pub token_id: TokenId,
+
+    /// The Ethererum token contract address.
+    #[serde(with = "serde_with::rust::display_fromstr")]
+    pub eth_token_contract_addr: EthAddr,
+
+    /// The auxiliary burn contract address (this is the contract that is used
+    /// in a Gnosis safe multi-sig withdrawal to record the matching TxOut
+    /// public key for the burn transaction on the MobileCoin blockchain).
+    #[serde(with = "serde_with::rust::display_fromstr")]
+    pub aux_burn_contract_addr: EthAddr,
+
+    /// The 4 bytes function signature that is used in the multi-sig
+    /// burn/withdrawal transaction. This is used as a sanity check.
+    pub aux_burn_function_sig: [u8; 4],
+}
+
+/// Configuration for a single safe we want to audit.
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
+pub struct AuditedSafeConfig {
+    /// The safe address.
+    #[serde(with = "serde_with::rust::display_fromstr")]
+    pub safe_addr: EthAddr,
+
+    /// The Gnosis safe transaction service API endpoint to sync from.
+    pub api_url: Url,
+
+    /// The tokens we want to audit.
+    pub tokens: Vec<AuditedToken>,
+}
+
+/// Configuration for Gnosis safe(s) auditing.
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
+pub struct GnosisSafeConfig {
+    /// The safe(s) we want to audit.
+    pub safes: Vec<AuditedSafeConfig>,
+}
+
+impl GnosisSafeConfig {
+    /// Load configuration data from a toml/json file.
+    pub fn load_from_path(path: impl AsRef<Path>) -> Result<Self, Error> {
```

Comment:
This is currently a fairly trivial method, but it's not tested.  
probably unlikely but what if the following was or gets mixed up
```rust
            Some("json") => Ok(toml::from_str(&data)?),
            Some("toml") => Ok(serde_json::from_str(&data)?),
```
If desired it might be good to update the test to use the `tempdir` crate or similar and write the test json and toml to files and then use `load_from_path()` function in the tests

something like, if PartialEq is added to the error the assert_matches could go away.
I'll reiterate the method is currently fairly trivial so can understand how this may feel like more work than benefit.
```rust
    static INPUT_TOML: &str = r#"
            [[safes]]
            safe_addr = "0x90213de428E9Ce4C77dD4943755Aa69cb2F803b7"
            api_url = "https://safe-api.example.com"

            [[safes.tokens]]
            token_id = 1
            eth_token_contract_addr = "0xd92e713d051c37ebb2561803a3b5fbabc4962431"
            aux_burn_contract_addr = "0x76BD419fBa96583d968b422D4f3CB2A70bf4CF40"
            aux_burn_function_sig = [0xc7, 0x6f, 0x06, 0x35]


            [[safes.tokens]]
            token_id = 2
            eth_token_contract_addr = "0x1111111111111111111111111111111111111111"
            aux_burn_contract_addr = "0x2222222222222222222222222222222222222222"
            aux_burn_function_sig = [0xaa, 0xbb, 0xcc, 0xdd]
       "#;

    static INPUT_JSON: &str = r#"{
            "safes": [
                {
                    "safe_addr": "0x90213de428E9Ce4C77dD4943755Aa69cb2F803b7",
                    "api_url": "https://safe-api.example.com",
                    "tokens": [
                        {
                            "token_id": 1,
                            "eth_token_contract_addr": "0xd92e713d051c37ebb2561803a3b5fbabc4962431",
                            "aux_burn_contract_addr": "0x76BD419fBa96583d968b422D4f3CB2A70bf4CF40",
                            "aux_burn_function_sig": [199, 111, 6, 53]
                        },
                        {
                            "token_id": 2,
                            "eth_token_contract_addr": "0x1111111111111111111111111111111111111111",
                            "aux_burn_contract_addr": "0x2222222222222222222222222222222222222222",
                            "aux_burn_function_sig": [170, 187, 204, 221]
                        }
                    ]
                }
            ]
        }"#;

    #[test]
    fn configs_from_path() {
        let dir = tempdir().unwrap();
        let toml_file = dir.path().join("testing.toml");
        fs::write(&toml_file, INPUT_TOML).unwrap();
        let json_file = dir.path().join("testing.json");
        fs::write(&json_file, INPUT_JSON).unwrap();
        let cfg1 = GnosisSafeConfig::load_from_path(toml_file).unwrap();
        let cfg2 = GnosisSafeConfig::load_from_path(json_file).unwrap();
        assert_eq!(cfg1, cfg2);
    }

    #[test]
    fn unsupported_extension() {
        let dir = tempdir().unwrap();
        let poml_file = dir.path().join("testing.poml");
        fs::write(&poml_file, INPUT_TOML).unwrap();
        let result = GnosisSafeConfig::load_from_path(poml_file);
        assert_matches!(result, Err(Error::UnrecognizedExtension(_)));
    }

    #[test]
    fn no_extension() {
        let dir = tempdir().unwrap();
        let file = dir.path().join("testing");
        fs::write(&file, INPUT_TOML).unwrap();
        let result = GnosisSafeConfig::load_from_path(file);
        assert_matches!(result, Err(Error::PathExtension));
    
```

Reply:
Personally I think that's overkill, we will find out very quickly if this was messed up. But I am fine adding all of this.
(Can't add Eq traits on `Error` because IoError doesn't implement them)